File: src/Server/Base/DateRange.cs
namespace Belin.Base;

using System.Globalization;
using System.Text.Json.Serialization;

/// <summary>
/// Defines a date range.
/// </summary>
/// <param name="Start">The start date.</param>
/// <param name="End">The end date.</param>
/// <param name="Type">The range type.</param>
public readonly record struct DateRange(DateTime Start, DateTime End, DateRangeType Type = DateRangeType.Custom): IEquatable<DateRange> {

	/// <summary>
	/// Creates a date range corresponding to the day including the specified date.
	/// </summary>
	/// <param name="date">The date.</param>
	/// <returns>The date range corresponding to the day including the specified date.</returns>
	public static DateRange Day(DateTime date) => new(
		date.Date,
		new DateTime(date.Year, date.Month, date.Day, 23, 59, 59, 999),
		DateRangeType.Day
	);

	/// <summary>
	/// Creates a date range corresponding to the month including the specified date.
	/// </summary>
	/// <param name="date">The date.</param>
	/// <returns>The date range corresponding to the month including the specified date.</returns>
	public static DateRange Month(DateTime date) => new(
		new DateTime(date.Year, date.Month, 1),
		new DateTime(date.Year, date.Month, DateTime.DaysInMonth(date.Year, date.Month), 23, 59, 59, 999),
		DateRangeType.Month
	);

	/// <summary>
	/// Creates a date range corresponding to the quarter including the specified date.
	/// </summary>
	/// <param name="date">The date.</param>
	/// <returns>The date range corresponding to the quarter including the specified date.</returns>
	public static DateRange Quarter(DateTime date) {
		var firstMonth = (date.GetQuarter() * 3) - 2;
		var lastMonth = firstMonth + 2;
		return new(
			new DateTime(date.Year, firstMonth, 1),
			new DateTime(date.Year, lastMonth, DateTime.DaysInMonth(date.Year, lastMonth), 23, 59, 59, 999),
			DateRangeType.Quarter
		);
	}

	/// <summary>
	/// Creates a date range corresponding to the week including the specified date.
	/// </summary>
	/// <param name="date">The date.</param>
	/// <returns>The date range corresponding to the week including the specified date.</returns>
	public static DateRange Week(DateTime date) {
		var delta = -(((int) date.DayOfWeek + 6) % 7);
		var sunday = date.AddDays(delta + 6);
		return new(
			date.AddDays(delta).Date,
			new DateTime(sunday.Year, sunday.Month, sunday.Day, 23, 59, 59, 999),
			DateRangeType.Week
		);
	}

	/// <summary>
	/// Creates a date range corresponding to the year including the specified date.
	/// </summary>
	/// <param name="date">The date.</param>
	/// <returns>The date range corresponding to the year including the specified date.</returns>
	public static DateRange Year(DateTime date) => new(
		new DateTime(date.Year, 1, 1),
		new DateTime(date.Year, 12, 31, 23, 59, 59, 999),
		DateRangeType.Year
	);

	/// <summary>
	/// Determines whether the specified object is equal to this object.
	/// </summary>
	/// <param name="other">An object to compare with this object.</param>
	/// <returns><see langword="true"/> if the specified object is equal to this object, otherwise <see langword="false"/>.</returns>
	public bool Equals(DateRange? other) => other is not null && Start == other.Value.Start && End == other.Value.End;

	/// <summary>
	/// Gets the label corresponding to this date range.
	/// </summary>
	/// <param name="culture">An object that supplies culture-specific formatting information.</param>
	/// <returns>The label corresponding to this date range.</returns>
	public string GetLabel(CultureInfo? culture = null) => Type switch {
		DateRangeType.Day => Start.ToString("d MMM yyyy", culture),
		DateRangeType.Week => Start.ToString($"S{Start.GetWeekOfYear(culture)} yyyy", culture),
		DateRangeType.Month => Start.ToString("MMMM yyyy", culture),
		DateRangeType.Quarter => Start.ToString($"T{Start.GetQuarter()} yyyy", culture),
		DateRangeType.Year => Start.Year.ToString(culture),
		_ => End.ToString("d MMM yyyy", culture)
	};
}

/// <summary>
/// Defines the type of a date range.
/// </summary>
[JsonConverter(typeof(JsonStringEnumConverter))]
public enum DateRangeType {

	/// <summary>
	/// A custom date range.
	/// </summary>
	Custom,

	/// <summary>
	/// A whole day.
	/// </summary>
	Day,

	/// <summary>
	/// A whole week.
	/// </summary>
	Week,

	/// <summary>
	/// A whole month.
	/// </summary>
	Month,

	/// <summary>
	/// A whole quarter.
	/// </summary>
	Quarter,

	/// <summary>
	/// A whole year.
	/// </summary>
	Year
}

File: src/Client/DateRange.ts
import {atMidnight, daysInMonth, getQuarter, getWeekOfYear} from "./Date.js";
import {TimeSpan} from "./TimeSpan.js";

/**
 * Defines the type of a date range.
 */
export const DateRangeType = Object.freeze({

	/**
	 * A custom date range.
	 */
	Custom: "Custom",

	/**
	 * A whole day.
	 */
	Day: "Day",

	/**
	 * A whole week.
	 */
	Week: "Week",

	/**
	 * A whole month.
	 */
	Month: "Month",

	/**
	 * A whole quarter.
	 */
	Quarter: "Quarter",

	/**
	 * A whole year.
	 */
	Year: "Year"
});

/**
 * Defines the type of a date range.
 */
export type DateRangeType = typeof DateRangeType[keyof typeof DateRangeType];

/**
 * Defines a date range.
 */
export class DateRange {

	/**
	 * The end date.
	 */
	readonly end: Date;

	/**
	 * The start date.
	 */
	readonly start: Date;

	/**
	 * The range type.
	 */
	readonly type: DateRangeType;

	/**
	 * Creates a new date range.
	 * @param start The start date.
	 * @param end The end date.
	 * @param type The range type.
	 */
	constructor(start: Date, end: Date, type: DateRangeType = DateRangeType.Custom) {
		this.start = start;
		this.end = end;
		this.type = type;
	}

	/**
	 * Creates a new date range from the specified JSON object.
	 * @param json A JSON object representing a date range.
	 * @returns The instance corresponding to the specified JSON object.
	 */
	static fromJson(json: Record<string, any>): DateRange {
		return new this(
			new Date(typeof json.start == "string" ? json.start : Date.now()),
			new Date(typeof json.end == "string" ? json.end : Date.now()),
			Object.values(DateRangeType).includes(json.type as DateRangeType) ? json.type as DateRangeType : DateRangeType.Custom
		);
	}

	/**
	 * Creates a date range corresponding to the day including the specified date.
	 * @param date The date.
	 * @returns The date range corresponding to the day including the specified date.
	 */
	static day(date: Date): DateRange {
		return new this(atMidnight(date), new Date(date.getFullYear(), date.getMonth(), date.getDate(), 23, 59, 59, 999), DateRangeType.Day);
	}

	/**
	 * Creates a date range corresponding to the month including the specified date.
	 * @param date The date.
	 * @returns The date range corresponding to the month including the specified date.
	 */
	static month(date: Date): DateRange {
		return new this(
			new Date(date.getFullYear(), date.getMonth(), 1),
			new Date(date.getFullYear(), date.getMonth(), daysInMonth(date), 23, 59, 59, 999),
			DateRangeType.Month
		);
	}

	/**
	 * Creates a date range corresponding to the quarter including the specified date.
	 * @param date The date.
	 * @returns The date range corresponding to the quarter including the specified date.
	 */
	static quarter(date: Date): DateRange {
		const firstMonth = (getQuarter(date) * 3) - 3;
		const lastMonth = firstMonth + 2;
		return new this(
			new Date(date.getFullYear(), firstMonth, 1),
			new Date(date.getFullYear(), lastMonth, daysInMonth(new Date(date.getFullYear(), lastMonth, 1)), 23, 59, 59, 999),
			DateRangeType.Quarter
		);
	}

	/**
	 * Creates a date range corresponding to the week including the specified date.
	 * @param date The date.
	 * @returns The date range corresponding to the week including the specified date.
	 */
	static week(date: Date): DateRange {
		const delta = -((date.getDay() + 6) % 7);
		const sunday = new Date(date.getTime() + ((delta + 6) * TimeSpan.MillisecondsPerDay));
		sunday.setHours(23, 59, 59, 999);
		return new this(atMidnight(new Date(date.getTime() + (delta * TimeSpan.MillisecondsPerDay))), sunday, DateRangeType.Week);
	}

	/**
	 * Creates a date range corresponding to the year including the specified date.
	 * @param date The date.
	 * @returns The date range corresponding to the year including the specified date.
	 */
	static year(date: Date): DateRange {
		return new this(new Date(date.getFullYear(), 0, 1), new Date(date.getFullYear(), 11, 31, 23, 59, 59, 999), DateRangeType.Year);
	}

	/**
	 * Returns a value indicating whether this date range is equal to the specified date range.
	 * @param other The other date range to compare.
	 * @returns `true` if this date range is equal to the specified date range, otherwise `false`.
	 */
	equals(other: DateRange): boolean {
		return this.start.getTime() == other.start.getTime() && this.end.getTime() == other.end.getTime();
	}

	/**
	 * The label corresponding to this date range.
	 * @param culture The current culture.
	 * @returns The label corresponding to this date range.
	 */
	getLabel(culture: Intl.Locale|string = navigator.language): string {
		const {start, end, type} = this;
		switch (type) {
			case DateRangeType.Day: return start.toLocaleString(culture, {dateStyle: "medium"});
			case DateRangeType.Week: return `S${getWeekOfYear(start)} ${start.getFullYear()}`;
			case DateRangeType.Month: return start.toLocaleString(culture, {month: "long", year: "numeric"});
			case DateRangeType.Quarter: return `T${getQuarter(start)} ${start.getFullYear()}`;
			case DateRangeType.Year: return start.getFullYear().toString();
			default: return end.toLocaleString(culture, {dateStyle: "medium"});
		}
	}
}
